src/encdatavault_fmt_plug.c

```diff
@@ -0,0 +1,407 @@
+	// ENCDataVault 1024 bits vault
+	{ "$encdv$3$4$bc93a92cf625e360$fe120bfb658d27b6$739f9d85964cef2f63b927ff77f3328cc5192014ae21c29954c322f4e808fe5c8abe64cc150dbcb08cb334f3b5c28357f10d8d5c6a103e2c899402136c14633aaf8a5347959b33b80ade2e3f5698864605940dc1423704999e5da859d6584491bbe940f00f162c75d2766b868ef2b4c6bf599a2e8ccea3f7cdfab193744a8d10", "123456789ABCDEf" },
+	{ NULL }
+};
+
+static int *cracked, cracked_count;
+static size_t cracked_size;
+
+static char (*saved_key)[PLAINTEXT_LENGTH + 1];
+
+static struct custom_salt {
+	unsigned int version;
+	unsigned int algo_id;
+	unsigned char salt[ENC_SALT_SIZE];
+	unsigned char iv[ENC_IV_SIZE];
+	unsigned char encrypted_data[ENC_BLOCK_SIZE];
+	unsigned char keychain[ENC_KEYCHAIN_SIZE];
+} *cur_salt;
+
+void enc_xor_block(uint64_t *dst, const uint64_t *src)
+{
+	dst[0] ^= src[0];
+	dst[1] ^= src[1];
+}
+
+void enc_aes_ctr_iterated(const unsigned char *in, unsigned char *out,
+                          const unsigned char *key, unsigned char ivs[7][16], size_t len,
+                          size_t nb_keys, uint64_t counter)
+{
+	AES_KEY aes_key;
+	uint64_t tmp_iv[2] = { 0 };
+	int i, j;
+
+	AES_set_encrypt_key(key, ENC_KEY_SIZE * 8, &aes_key);
+	len = len >> 4;
+	counter = counter << 56;
+
+	for (i = 0; i < len; i++) {
+		tmp_iv[0] = ((uint64_t *) ivs[0])[0];
+		tmp_iv[1] = counter;
+		AES_encrypt((unsigned char *)tmp_iv, out, &aes_key);
+		enc_xor_block((uint64_t *) out, (uint64_t *) in);
+
+		for (j = 1; j < nb_keys; j++) {
+			tmp_iv[0] = ((uint64_t *)(ivs[j]))[0];
+			tmp_iv[1] = counter;
+			AES_encrypt((unsigned char *)tmp_iv, (unsigned char *)tmp_iv,
+			            &aes_key);
+			enc_xor_block((uint64_t *) out, tmp_iv);
+		}
+		// Increment counter, only 255 block encryption is supported.
+		counter += 0x0100000000000000;
+		out += AES_BLOCK_SIZE;
+		in += AES_BLOCK_SIZE;
+	}
+}
+
+static void init(struct fmt_main *self)
+{
+	omp_autotune(self, OMP_SCALE);
+	saved_key = mem_calloc_align(sizeof(*saved_key),
+	                             self->params.max_keys_per_crypt, MEM_ALIGN_WORD);
+	cracked = mem_calloc(sizeof(*cracked), self->params.max_keys_per_crypt);
+	cracked_size = sizeof(*cracked) * self->params.max_keys_per_crypt;
+	cracked_count = self->params.max_keys_per_crypt;
+}
+
+static void done(void)
+{
+	MEM_FREE(saved_key);
+	MEM_FREE(cracked);
+}
+
+static int valid(char *ciphertext, struct fmt_main *self)
+{
+	char *p = ciphertext, *ctcopy, *keeptr;
+	int extra;
+
+	if (strncmp(ciphertext, FORMAT_TAG, FORMAT_TAG_LENGTH) != 0)
+		return 0;
+
+	ctcopy = strdup(ciphertext);
+	keeptr = ctcopy;
+	ctcopy += FORMAT_TAG_LENGTH;
+	if ((p = strtokm(ctcopy, "$")) == NULL) // version
+		goto err;
+	if (!isdec(p))
+		goto err;
+	if ((p = strtokm(NULL, "$")) == NULL)   // algorithm id
+		goto err;
+	if (!isdec(p))
+		goto err;
+	if (atoi(p) > 4 || atoi(p) < 1) {
+		fprintf(stderr, "Warning: algorithm id %d not supported!\n", atoi(p));
+		goto err;
+	}
```

Comment:
So we accept version 1, 2 and 3 but I only see tests for 1 or `else` in `crypt_all`. Is there a version 2 and does this code crack it too? If no to any of those, we should also reject version 2 here.

Reply:
You are right but the check needs to be done on the version number. I corrected it.

Reply:
BTW we usually avoid printing warnings from `valid()` becuase it doesn't go well with auto-selection of format, but since this format has its own tag they won't hurt much. What we usually do in such cases though is something like:
```c
if (atoi(p) != 1 && atoi(p) != 3) {
	static int warning;
	if (!warning++)
 		fprintf(stderr, "%s: Warning: version %d not supported, not loading such hashes!\n", atoi(p), self->params.label);
	goto err;
}
```
another possibility is only showing warnings if verbosity was bumped:
```c
	if (options.verbosity >= VERB_DEBUG)
 		fprintf(stderr, "%s: Warning: version %d not supported, not loading such hashes!\n", atoi(p), self->params.label);
```
Perhaps @solardiz have some preference here?

Reply:
May be the version should be reported as a tunable cost.

Reply:
> May be the version should be reported as a tunable cost.

Commenting-out test vectors, it seems the shorter v1 hashes do about 65Kp/s while the longer v3 hash does 71K. I would have guessed the other way round.

Reply:
I have not a big difference and the main time is spend in the 1000 MD5 iterations and it should not have big difference for AES decryption.